### PR TITLE
Add the new scanner option "time_between_request". This option avoid …

### PR DESCRIPTION
**Pull request by jjnicola**
…quick connection establishment, termination and quick request to a single host.

* nasl/nasl_socket.c: Include gvm/base/prefs.h, stdlib.h and sys/time.h.
  (wait_before_next_probe): New function.
  (nasl_open_privileged_socket, nasl_open_sock_tcp_bufsz, nasl_send,
  nasl_close_socket): Call new function.

* src/openvassd.c (openvassd_defaults): Add new option time_between_request.

* doc/openvassd.8.in: Add description for the new option time_between_request.

### CODE REVIEW

File: nasl/nasl_socket.c [hunk 1]
```diff
@@ -109,6 +113,50 @@ block_socket (int soc)
   return 0;
 }
 
+static void
+wait_before_next_probe ()
+{
+
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Fixed

File: nasl/nasl_socket.c [hunk 2]
```diff
@@ -109,6 +113,50 @@ block_socket (int soc)
...
+  if (minwaittime > 0)
+    {
+      static double lastprobesec = 0;
+      static double lastprobeusec = 0;
+      struct timeval tvnow, tvdiff;
+      double diff_msec;
+      int time2wait = 0;
+
+      gettimeofday(&tvnow, NULL);
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Fixed

File: nasl/nasl_socket.c [hunk 3]
```diff
@@ -109,6 +113,50 @@ block_socket (int soc)
...
+      if (lastprobesec <= 0)
+        {
+         lastprobesec = tvnow.tv_sec - 10;
+         lastprobeusec = tvnow.tv_usec;
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Fixed

File: nasl/nasl_socket.c [hunk 4]
```diff
@@ -109,6 +113,50 @@ block_socket (int soc)
...
+      tvdiff.tv_sec = tvnow.tv_sec - lastprobesec;
+      tvdiff.tv_usec = tvnow.tv_usec - lastprobeusec;
+      if (tvdiff.tv_usec <= 0)
+        {
+          tvdiff.tv_sec += 1;
+          tvdiff.tv_usec *= -1;
+        }
+
+      diff_msec = tvdiff.tv_sec * 1000 + tvdiff.tv_usec / 1000;
+      time2wait  = (minwaittime - diff_msec)*1000;
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Fixed

File: nasl/nasl_socket.c [hunk 5]
```diff
@@ -218,6 +266,8 @@ nasl_open_privileged_socket (lex_ctxt * lexic, int proto)
 
 
 restart:
+  if (proto == IPPROTO_TCP)
+    wait_before_next_probe ();
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Regarding opening and closing sockets, I think it is needed based on this example:
https://support.industry.siemens.com/cs/document/28970909/cp-443-1-firmware-update-v2-7-and-unrestricted-use-on-every-hardware-release-(6gk7443-1ex10-0xe0-and-6gk7443-1ex11-0xe0)?dti=0&lc=en-WW

where can be read that for this device "In the case of quick connection establishment and termination scenarios via the TCP/IP protocol, the TCP/IP sockets could be permanently occupied. As a result, certain communication connections could not be established anymore."

File: nasl/nasl_socket.c [hunk 6]
```diff
@@ -907,8 +961,10 @@ nasl_close_socket (lex_ctxt * lexic)
 
   soc = get_int_var_by_num (lexic, 0, -1);
   if (fd_is_stream (soc))
-    return close_stream_connection (soc) < 0 ? NULL : FAKE_CELL;
-
+    {
+      wait_before_next_probe ();
+      return close_stream_connection (soc) < 0 ? NULL : FAKE_CELL;
+    }
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Same.

File: nasl/nasl_socket.c [hunk 7]
```diff
@@ -417,6 +467,8 @@ nasl_open_sock_tcp_bufsz (lex_ctxt * lexic, int bufsz)
   if (port < 0)
     return NULL;
 
+  wait_before_next_probe ();
+
```
[the remark this answers is not in the document]
Comment (jjnicola) [reply nasl/nasl_socket.c]: Same

File: nasl/nasl_socket.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (kroosec) [reply nasl/nasl_socket.c]: > Some NVTs has this behavior (quick open/close calls in row). Also we don't have control of the time between two NVT are launched. I think that a waiting time before opening a socket could be helpful in this case.

Even in such a case, it seems that at least the delay before closing is still unneeded.

Tagging @cfi-gb for extra input.

[the remark this answers is not in the document]
Comment (cfi-gb) [reply nasl/nasl_socket.c]: I'm quite undecided about this as we're currently don't know how similar devices are behaving.

In this specific scenario i was able to work around the mentioned issues by putting a sleep(2); before each open socket call but different devices might require a wait before closing the socket as well.

**Pull request merged**